### DISCUSSION TITLE
Consul services updated realtime in background

### DISCUSSION THREAD

**Discussion by daveblack93** [ThreeMammals/Ocelot#1910] [Ideas]
Hello,
We are using Hashicorp Consul in order to communicate between microservices and Ocelot to route api call from Frontend webpages to api services. Till now our api services where deployed on "static" VMs with a fixed dns/ip, so at api service startup we retrieve our third parties services URL and store them into a dictionary and this was enough until the end of life of the microservice.
Some weeks ago, We started migrating microservice from "static" IIS VM deploy, to Hashicorp Nomad cluster.
Now we need to keep updated the third parties services URL real-time because our services are deployed in a cluster of VMs, so we don't know every deploy where the orchestrator will put the service and the corresponding URL.

Now, after some scouting, we found that Consul Service catalogue support Blocking Queries (based on long polling architecture) in order to keep update third parties services URL without calling every time Consul service catalogue or polling with a delay of X seconds. (https://developer.hashicorp.com/consul/api-docs/features/blocking)

My question here is: do you think this could be a new GlobalConfiguration.GlobalConfiguration.Type useful in your project?

Benefits are that the Service URL is kept updated real-time without overloading Consul service Catalogue and without configure a PollInterval.

Thanks,
Dave.

**Comment by raman-m** [comment]
Hi Dave!
Welcome to Ocelot world! 🐯 

---
> Some weeks ago, We started migrating microservice from "static" IIS VM deploy, to Hashicorp Nomad cluster.

Lol Good luck!  Do you have DevOps engineer? 😉 

----
> My question here is: do you think this could be a new GlobalConfiguration.GlobalConfiguration.Type useful in your project?

Sorry? Didn't get it! Probably you've meant `GlobalConfiguration.ServiceDiscoveryProvider.Type`?! Or?... Why do you need new type of service discovery? What's the goal?

-----
> Benefits are that the Service URL is kept updated real-time without overloading Consul service Catalogue and without configure a PollInterval.

Cannot get it... What's wrong here? 
Ocelot pools the services based on "**PollingInterval**" property.
Consul is responsible for keeping all services configs. Why do you think that Ocelot should manage Consul Catalogue or moreover Consul cluster?

@ggnaegi Correct me if I'm wrong please, or you will have better understanding of this user scenario with Consul.

**Reply by ggnaegi** [reply]
@raman-m, @daveblack93  It's very interesting indeed. It means we wouldn't need a polling interval anymore, eg. the system waits for changes and then updates the client list.

_On subsequent requests for this resource, the client can set the index query string parameter to the value of X-Consul-Index, indicating that the client wishes to wait for any changes subsequent to that index._

_When this is provided, the HTTP request will "hang" until a change in the system occurs, or the maximum timeout is reached. A critical note is that the return of a blocking request is no guarantee of a change. It is possible that the timeout was reached or that there was an idempotent write that does not affect the result of the query._

A bit of refactoring is needed but it's worth a try. @daveblack93 could you at least create an issue, describing the use case:
eg. "Ocelot should be able to obtain the current list of services, without this involving short polling intervals or updating the list of services for each request."

**Reply by raman-m** [reply]
What I've got is that Mr. Dave doesn't like "**PollingInterval**" property and doesn't want to manage it.
Also, keep in mind we cannot remove "**PollingInterval**" property because of backward compatibility.
Also, "**the system waits for changes**" means Ocelot provider subscribes to Consul notification events. Otherwise I don't understand the solution with Timeout management which will consume a significant value of operation memory because Http requests should wait a response from Consul for unknown time period, so they stay in memory.
Current `Timer` callback object consumes less memory I think when polling services updates each PollingInterval.

----
> On subsequent requests for this resource, the client can set the index query string parameter to the value of `X-Consul-Index`, indicating that the client wishes to wait for any changes subsequent to that index.

This approach requires to cache or keep & manage `X-Consul-Index` objects in-memory.

**Reply by raman-m** [reply]
The quote from [Blocking Queries](https://developer.hashicorp.com/consul/api-docs/features/blocking):

> In addition to `index`, endpoints that support blocking will also honor a `wait` parameter specifying a maximum duration for the blocking request. This is limited to 10 minutes. If not set, the wait time defaults to 5 minutes. This value can be specified in the form of "10s" or "5m" (i.e., 10 seconds or 5 minutes, respectively). A small random amount of additional wait time is added to the supplied maximum `wait` time to spread out the wake up time of any concurrent requests. This adds up to `wait / 16` additional time to the maximum duration.

I don't see a difference with current "**PollingInterval**" functionality. Do you see.
Seems "PollingInterval" does exactly the same.

**Reply by daveblack93** [reply]
Hi @raman-m,
the idea that I have is the following:
- **PollingInterval**: Ocelot will poll consul every x milliseconds defined in configuration to get the state of the art Service Catalogue configuration.
- **Blocking Query**: Ocelot will call catalogue to retrieve the Address:Port of the services at startup and will store for every service the configuration and the index. Then will open in parallel n tasks as many service you have configured, implementing an infinite loop that wait if some changes occurred in services; If some service will change index, Consul will get the latest information (Address:Port & X-Consul-Index) update the Ocelot configuration and re-wait for changes.
For sure if nothing happen, Consul after a default timeout will reply and a new LongPolling will be initialize.

For my understanding these two scenarios (Polling vs Blocking) are quite different... Correct me if I'm wrong ;).

I will provide some code snippet example:
```c#
Parallel.ForEach(_externalServiceSettingList,
            serviceSetting =>
            {
                //We keep updated service with long polling
                myConsulService.KeepSyncExternalService(serviceSetting, stoppingToken);
            });

private async Task KeepSyncExternalService(ExternalServiceSetting setting, CancellationToken ctx)
{
do
        {
            var serviceName = setting.ServiceName;
            var options = new QueryOptions() { WaitIndex = setting.LastIndex };
            var onlyPassingService = true;
            var serviceResult = await _consulClient.Health.Service(serviceName, string.Empty, onlyPassingService, options, ctx)
                       .ConfigureAwait(false);
            setting.LastIndex = serviceResult.LastIndex;
        } while (true);
}
```

**Reply by ggnaegi** [reply]
@daveblack93 Is it your code or chat GPT? 😸 Well, it means we need somewhere a background task calling consul to retrieve the service list, probably a custom consul client service that is then injected to the provider. It would work like that, but it means we would get rid of the polling mechanism. I don't feel so sad about this.

**Reply by daveblack93** [reply]
Hi @ggnaegi, is not gpt-code, is code frag coming from my custom library implementation.
As you mentioned, I initialize a backgroundWorker with the logic described.
Regarding polling mechanism, honestly I would say that my proposal was a possible additional feature but if you say that is better get rid of that... you have more experience than me ;)

**Reply by ggnaegi** [reply]
@daveblack93 The polling has been introduced to address a concern, performance issues when calling consul for each request. In my case, the environment is stable, polling consul every 4 hours is fine. But, in your case, you don't know when the service configurations are going to change. So, it means, you will use a short polling interval. It's incredibly chatty! For your use case, the long polling is the only working solution imo. We should think about it...

**Reply by ggnaegi** [reply]
@daveblack93 @raman-m  I would propose the following solution: Add a new provider for consul long polling named "LongPolling" and keep "PollConsul" for use cases like mine (in my case the long polling would be too chatty) or for backward compatibility. 
The design should be:
- A new IServiceDiscoveryProvider implementation with a background task running during provider lifetime. So we can keep the "per request" logic from Ocelot, but still have an implicit singleton.
- GetAsync: if nodes list is null, we retrieve the data from consul, with headers etc, update the nodes list, the id list  and we start  or restart the background task for the current service name.
- GetAsync: if nodes list is not null, we return the nodes list

Why such implementation?
Yes, the first request to an endpoint of the given service will last longer, but we can then keep the ability to modify the ocelot configuration file stored in Consul KV-Store on the fly. We avoid having a global configuration too. Imo it's a good compromise, the changes are not going to be dramatic for a first version.

**Reply by raman-m** [reply]
@ggnaegi 
We need to think in terms of Ocelot Health and Microservices readiness. So SD routes must be ready after Ocelot app start. But they are not now... Only background workers can help.
So waiting for the first SD request is bad because as you said first request will be very slow and users will see spikes in logs.

-----
> We avoid having a global configuration too.

Unclear for me... Seems we have to define Service Discovery Provider **type** always in global config.

**Reply by ggnaegi** [reply]
@raman-m we are talking about one request, come on. So, maybe 100-200 ms overhead for one request. And even, for one request for the given service name... If it lasts more than 200 ms to retrieve the client list, then you might have an infrastructure problem somewhere...

**Reply by ggnaegi** [reply]
@daveblack93 we could have something like this... But we will need several weeks to test and validate.
```c#
using System.Collections.Immutable;
using Ocelot.Infrastructure.Extensions;
using Ocelot.Logging;
using Ocelot.ServiceDiscovery.Providers;
using Ocelot.Values;
using System.Threading;

namespace Ocelot.Provider.Consul;

public class Consul : IServiceDiscoveryProvider
{
    private const string VersionPrefix = "version-";
    private readonly ConsulRegistryConfiguration _config;
    private readonly IConsulClient _consul;
    private readonly IOcelotLogger _logger;
    private readonly SemaphoreSlim _semaphoreSlim = new(1, 1);
    private ImmutableList<Service> _services;
    private readonly ConsulProviderOptions _consulProviderOptions;

    public Consul(ConsulRegistryConfiguration config, IOcelotLoggerFactory factory, IConsulClientFactory clientFactory, ConsulProviderOptions consulProviderOptions)
    {
        _config = config;
        _consul = clientFactory.Get(_config);
        _logger = factory.CreateLogger<Consul>();
        _consulProviderOptions = consulProviderOptions;
    }

    public async Task<List<Service>> GetAsync()
    {
        await _semaphoreSlim.WaitAsync();
        try
        {
            if (_services != null)
            {
                return[.. _services];
            }

            (_services, var initialWaitIndex) = await RetrieveServiceListFromConsulAsync(0);

            ConfigurePollingType(initialWaitIndex);
            
            return[.. _services];
        }
        finally
        {
            _semaphoreSlim.Release();
        }
    }

    public string ServiceName => _consulProviderOptions.ServiceName;

    private void ConfigurePollingType(ulong initialWaitIndex)
    {
        if (_consulProviderOptions.PollingType == ConsulPollingType.None)
        {
            return;
        }

        if (_consulProviderOptions.PollingType == ConsulPollingType.LongPolling)
        {
            _ = LongPollingServiceListAsync(initialWaitIndex);
            return;
        }

        _ = new Timer(_ =>
        {
            (_services, var waitIndex) = RetrieveServiceListFromConsulAsync(0).Result;
        }, null, TimeSpan.Zero, TimeSpan.FromMilliseconds(_consulProviderOptions.PollingInterval));
    }

    private async Task LongPollingServiceListAsync(ulong initialWaitIndex)
    {
        var waitIndex = initialWaitIndex;
        while (true)
        {
            try
            {
                (_services, waitIndex) = await RetrieveServiceListFromConsulAsync(waitIndex);
            }
            catch (Exception ex)
            {
                _logger.LogError(() => $"Error in long polling: {ex.Message}", ex);
            }
        }
    }

    public virtual async Task<(ImmutableList<Service> ServiceList, ulong WaitIndex)> RetrieveServiceListFromConsulAsync(ulong waitIndex)
    {
        var queryResult = await _consul.Health.Service(_config.KeyOfServiceInConsul, string.Empty, true, new QueryOptions
        {
            WaitIndex = waitIndex,
        });

        var services = new List<Service>();
        foreach (var serviceEntry in queryResult.Response)
        {
            var service = serviceEntry.Service;
            if (IsValid(service))
            {
                var nodes = await _consul.Catalog.Nodes();
                if (nodes.Response == null)
                {
                    services.Add(BuildService(serviceEntry, null));
                }
                else
                {
                    var serviceNode = nodes.Response.FirstOrDefault(n => n.Address == service.Address);
                    services.Add(BuildService(serviceEntry, serviceNode));
                }
            }
            else
            {
                _logger.LogWarning(
                    () => $"Unable to use service address: '{service.Address}' and port: {service.Port} as it is invalid for the service: '{service.Service}'. Address must contain host only e.g. 'localhost', and port must be greater than 0.");
            }
        }

        return (services.ToImmutableList(), queryResult.LastIndex);
    }

    private static Service BuildService(ServiceEntry serviceEntry, Node serviceNode)
    {
        var service = serviceEntry.Service;
        return new Service(
            service.Service,
            new ServiceHostAndPort(
                serviceNode == null ? service.Address : serviceNode.Name,
                service.Port),
            service.ID,
            GetVersionFromStrings(service.Tags),
            service.Tags ?? Enumerable.Empty<string>());
    }

    private static bool IsValid(AgentService service)
        => !string.IsNullOrEmpty(service.Address)
        && !service.Address.Contains($"{Uri.UriSchemeHttp}://")
        && !service.Address.Contains($"{Uri.UriSchemeHttps}://")
        && service.Port > 0;

    private static string GetVersionFromStrings(IEnumerable<string> strings)
        => strings?.FirstOrDefault(x => x.StartsWith(VersionPrefix, StringComparison.Ordinal))
            .TrimStart(VersionPrefix);
}
```

**Reply by daveblack93** [reply]
@ggnaegi this is exactly what I was thinking. Take your time to validate it ;)

**Reply by raman-m** [reply]
@ggnaegi  Good but draft code 💪 Need to convert this discussion to an issue?
What's your motivation in delivery of this idea-feature?

**Reply by ggnaegi** [reply]
@raman-m Yes, it's a draft, but we would cover all use cases for the consul service discovery: The ones with a stable environment (Polling), the others with a more dynamic and uncertain environment (Long Polling)...

**Reply by raman-m** [reply]
@ggnaegi  Seems the author will not help you...
In what release will you work on that? And when would you like to start?

#### P.S.
We have big Annual 2023 release starting soon... Really large scope!

**Reply by daveblack93** [reply]
Hi @raman-m sorry if I do not reply to your message. I could help you and @ggnaegi developing this feature. I have to know a bit better how Ocelot and Consul interact one another. But sure I can help you :)

**Reply by ggnaegi** [reply]
@daveblack93 Unfortunately, I will not have much time during the next two weeks. So, do you know xunit? Could we agree on something, I create the issue, create the PR, authorize you on my forked repo and then you work on the unit and acceptance tests? I'm curious, where are you from? I will explain you, where the code has been modified and why.

**Reply by daveblack93** [reply]
I'm from Italy ;) yes sure we could do something like that. I will find some times in my free time ;)

**Reply by raman-m** [reply]
@ggnaegi  I believe @daveblack93 can fork Ocelot to his account and invite you.
@daveblack93 Please fork the repo!

**Reply by ggnaegi** [reply]
@raman-m I prefer to go the way I described, it's easier for me.

**Comment by ggnaegi** [comment]
@daveblack93 Great idea!

**Reply by raman-m** [reply]
OK, maybe it is great.
Why not to review our current design where we refresh the services list by [GetAsync()](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot.Provider.Consul/PollConsul.cs#L39) method being called by [ILoadBalancerCreator objects](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/DependencyInjection/OcelotBuilder.cs#L88-L91) in `Create` method and appropriate LoadBalancer constructors have no injectable scoped services except maybe attached `Func` delegates.
